go/viperutil/internal/sync/sync_test.go

```diff
@@ -0,0 +1,258 @@
+limitations under the License.
+*/
+
+package sync_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"os"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/spf13/viper"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"vitess.io/vitess/go/viperutil"
+	vipersync "vitess.io/vitess/go/viperutil/internal/sync"
+	"vitess.io/vitess/go/viperutil/internal/value"
+)
+
+func TestPersistConfig(t *testing.T) {
+	type config struct {
+		Foo int `json:"foo"`
+	}
+
+	loadConfig := func(t *testing.T, f *os.File) config {
+		t.Helper()
+
+		data, err := os.ReadFile(f.Name())
+		require.NoError(t, err)
+
+		var cfg config
+		require.NoError(t, json.Unmarshal(data, &cfg))
+
+		return cfg
+	}
+
+	setup := func(t *testing.T, v *vipersync.Viper, minWaitInterval time.Duration) (*os.File, chan struct{}) {
+		tmp, err := os.CreateTemp(t.TempDir(), fmt.Sprintf("%s_*.json", strings.ReplaceAll(t.Name(), "/", "_")))
+		require.NoError(t, err)
+
+		t.Cleanup(func() { os.Remove(tmp.Name()) })
+
+		cfg := config{
+			Foo: jitter(1, 100),
+		}
+
+		data, err := json.Marshal(&cfg)
+		require.NoError(t, err)
+
+		_, err = tmp.Write(data)
+		require.NoError(t, err)
+
+		static := viper.New()
+		static.SetConfigFile(tmp.Name())
+		require.NoError(t, static.ReadInConfig())
+
+		ch := make(chan struct{}, 1)
+		v.Notify(ch)
+
+		cancel, err := v.Watch(context.Background(), static, minWaitInterval)
+		require.NoError(t, err)
+		t.Cleanup(cancel)
+
+		return tmp, ch
+	}
+
+	t.Run("basic", func(t *testing.T) {
+		v := vipersync.New()
+
+		minPersistWaitInterval := 2 * time.Second
+		get := vipersync.AdaptGetter("foo", viperutil.GetFuncForType[int](), v)
+		f, ch := setup(t, v, minPersistWaitInterval)
+
+		old := get("foo")
+		loadConfig(t, f)
+		v.Set("foo", old+1)
+		// This should happen immediately in-memory and on-disk.
+		assert.Equal(t, old+1, get("foo"))
+		<-ch
+		assert.Equal(t, old+1, loadConfig(t, f).Foo)
+
+		v.Set("foo", old+2)
+		// This should _also_ happen immediately in-memory, but not on-disk.
+		// It will take up to 2 * minPersistWaitInterval to reach the disk.
+		assert.Equal(t, old+2, get("foo"))
+		assert.Equal(t, old+1, loadConfig(t, f).Foo)
+
+		select {
+		case <-ch:
+		case <-time.After(2 * minPersistWaitInterval):
+			assert.Fail(t, "config was not persisted quickly enough", "config took longer than %s to persist (minPersistWaitInterval = %s)", 2*minPersistWaitInterval, minPersistWaitInterval)
+		}
+
+		assert.Equal(t, old+2, loadConfig(t, f).Foo)
+	})
+
+	t.Run("no persist interval", func(t *testing.T) {
+		v := vipersync.New()
+
+		var minPersistWaitInterval time.Duration
+		get := vipersync.AdaptGetter("foo", viperutil.GetFuncForType[int](), v)
+		f, ch := setup(t, v, minPersistWaitInterval)
+
+		old := get("foo")
+		loadConfig(t, f)
+		v.Set("foo", old+1)
+		// This should happen immediately in-memory and on-disk.
```

Comment:
Have we run these tests e.g. 20 times locally and maybe 10 times in the CI? These sorts of assumptions make me nervous given how unpredictable our CI env has been. :-)

Reply:
it's not an assumption, it's an assertion (but duplicated in comment form). if this is not true, the test should fail (and yes i've run them a bunch locally, we'll see how CI does as i push up more changes for your other feedback!)